src/main/resources/org/openjdk/jextract/impl/resources/RuntimeHelper.java.template

```diff
@@ -99,147 +110,60 @@ final class RuntimeHelper {
-            }
-            mtype = mtype.appendParameterTypes(Object[].class);
-            boolean needsAllocator = function.returnLayout().isPresent() &&
-                                function.returnLayout().get() instanceof GroupLayout;
-            if (needsAllocator) {
-                mtype = mtype.insertParameterTypes(0, SegmentAllocator.class);
-            } else {
-                handle = MethodHandles.insertArguments(handle, 0, THROWING_ALLOCATOR);
-            }
-            return handle.asType(mtype);
-        }
-
-        static Class<?> carrier(MemoryLayout layout, boolean ret) {
-            if (layout instanceof ValueLayout valueLayout) {
-                return valueLayout.carrier();
-            } else if (layout instanceof GroupLayout) {
-                return MemorySegment.class;
-            } else {
-                throw new AssertionError("Cannot get here!");
-            }
-        }
-
-        private Object invoke(SegmentAllocator allocator, Object[] args) throws Throwable {
-            // one trailing Object[]
-            int nNamedArgs = function.argumentLayouts().size();
-            assert(args.length == nNamedArgs + 1);
-            // The last argument is the array of vararg collector
-            Object[] unnamedArgs = (Object[]) args[args.length - 1];
-
-            int argsCount = nNamedArgs + unnamedArgs.length;
-            Class<?>[] argTypes = new Class<?>[argsCount];
-            MemoryLayout[] argLayouts = new MemoryLayout[nNamedArgs + unnamedArgs.length];
-
-            int pos = 0;
-            for (pos = 0; pos < nNamedArgs; pos++) {
-                argLayouts[pos] = function.argumentLayouts().get(pos);
-            }
-
-            assert pos == nNamedArgs;
-            for (Object o: unnamedArgs) {
-                argLayouts[pos] = variadicLayout(normalize(o.getClass()));
-                pos++;
-            }
-            assert pos == argsCount;
-
-            FunctionDescriptor f = (function.returnLayout().isEmpty()) ?
-                    FunctionDescriptor.ofVoid(argLayouts) :
-                    FunctionDescriptor.of(function.returnLayout().get(), argLayouts);
-            MethodHandle mh = LINKER.downcallHandle(symbol, f);
-            boolean needsAllocator = function.returnLayout().isPresent() &&
-                                            function.returnLayout().get() instanceof GroupLayout;
-            if (needsAllocator) {
-                mh = mh.bindTo(allocator);
-            }
-            // flatten argument list so that it can be passed to an asSpreader MH
-            Object[] allArgs = new Object[nNamedArgs + unnamedArgs.length];
-            System.arraycopy(args, 0, allArgs, 0, nNamedArgs);
-            System.arraycopy(unnamedArgs, 0, allArgs, nNamedArgs, unnamedArgs.length);
-
-            return mh.asSpreader(Object[].class, argsCount).invoke(allArgs);
+    private static Class<?> unboxIfNeeded(Class<?> clazz) {
```

Comment:
question - wouldn't it be simpler to have a single switch from carrier type to variadic layout directly? I think the current code organization is the way it is because we used to need both carriers and function descriptor, but now we only need the latter.

Reply:
Thanks for the tip. I've managed to simplify this code quite a bit: https://github.com/openjdk/jextract/pull/149/commits/a8c387c89a4ea00dabbbc50479994c5e6b4714a4

- We only care about wrapper types, not primitves, since all the arguments come to us in an `Object[]`.
- Smaller-than-int types should be promoted to `int`, not `long`.
- No need to handle `void` (in fact, that would be wrong).